File: java/LamportClock.java
import java.net.*;
import java.util.Scanner;
import java.util.Map;
import java.util.concurrent.locks.ReentrantLock;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

class LamportClock {
    private int clock;
    private ReentrantLock threadLock = new ReentrantLock();
    private static Map<Integer, Integer> PROCESS_PORT = Map.of(
        1, 8080,
        2, 8081,
        3, 8082
    );

    LamportClock() {
        this.clock = 0;
    }

    void receiveThread(int processNumber) {
        int processPort = PROCESS_PORT.get(processNumber);
        try {
            ServerSocket serverSocket = new ServerSocket(processPort);
            while (true) {
                Socket socket = serverSocket.accept();
                DataInputStream input = new DataInputStream(socket.getInputStream());
                int destClock = input.readInt();
                socket.close();
                gotMessage(destClock);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    void calcEvent() {
        threadLock.lock();
        this.clock++;
        System.out.println("時刻" + this.clock + ": 計算イベント");
        threadLock.unlock();
    }

    void gotMessage(int anotherProcessClock) {
        threadLock.lock();
        this.clock = Math.max(anotherProcessClock, this.clock) + 1;
        System.out.println("時刻" + this.clock + ": メッセージを受信");
        threadLock.unlock();
    }

    void sendMessage(int destProcess) {
        int destPort = PROCESS_PORT.get(destProcess);
        try {
            Socket socket = new Socket("127.0.0.1", destPort);
            DataOutputStream output = new DataOutputStream(socket.getOutputStream());
            threadLock.lock();
            this.clock++;
            System.out.println("時刻" + this.clock + ": メッセージを送信");
            output.writeInt(this.clock);
            socket.close();
            threadLock.unlock();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    void showTime() {
        threadLock.lock();
        System.out.println("現在の時刻: " + this.clock);
        threadLock.unlock();
    }

    public static void main(String[] args) {

            int processNumber = Integer.parseInt(args[0]);
            LamportClock lamportClock = new LamportClock();
            Thread receiveThread = new Thread(() -> lamportClock.receiveThread(processNumber));
            Scanner scanner = new Scanner(System.in);
            receiveThread.start();
            while (true) {
                System.out.println("\n=== 時刻" + lamportClock.clock + " ===");
                System.out.println("イベントの種類を決めてください(c: 計算イベント, s: メッセージ送信, t: 現在の時刻を表示): ");
                String cmd = scanner.nextLine();
                switch (cmd) {
                    case "c":
                        lamportClock.calcEvent();
                        break;
                    case "s":
                        System.out.print("プロセスナンバー: ");
                        int processNum = Integer.parseInt(scanner.nextLine());
                        lamportClock.sendMessage(processNum);
                        break;
                    case "t":
                        lamportClock.showTime();
                        break;
                    default:
                        System.out.println("再入力");
                        break;
                }
            }
       
    }
}

File: logical_clock/js/lamportclock.js
const net = require('net');
const readline = require('readline');
const { Mutex } = require('async-mutex');

const PROCESS_PORT = {//OK
    1: 8080,
    2: 8081,
    3: 8082,
};

class LamportClock {//OK
    constructor() {
        this.clock = 0;
        this.mutex = new Mutex();
    }

    async receiveThread(processNumber) {//非同期関数として定義
        const processPort = PROCESS_PORT[processNumber];

        const server = net.createServer(async (socket) => {
            const destRawClock = await this.receiveMessage(socket);
            const destClock = parseInt(destRawClock, 10);
            await this.gotMessage(destClock);
        });

        server.listen(processPort, '127.0.0.1');
    }

    async receiveMessage(socket) {
        return new Promise((resolve) => {
            socket.on('data', (data) => {
                resolve(data.toString());
                socket.destroy();
            });
        });
    }

    async calcEvent() {
        const release = await this.mutex.acquire();//ロック取得
        try {
            console.log(`時刻${this.clock}: 計算イベント`);
            this.clock += 1;
        } finally {
            release();
        }
    }

    async gotMessage(anotherProcessClock) {
        const release = await this.mutex.acquire();
        try {
            const l = Math.max(this.clock, anotherProcessClock);
            this.clock = l + 1;
            console.log(`時刻${this.clock}: メッセージを受信`);
        } finally {
            release();
        }
    }

    async sendMessage(destProcess) {
        const destPort = PROCESS_PORT[destProcess];
        const client = new net.Socket();
        const release = await this.mutex.acquire();
        try {
            this.clock += 1;
            console.log(`時刻${this.clock}: メッセージを送信`);
            client.connect(destPort, '127.0.0.1', () => {
                client.write(this.clock.toString());
                client.destroy();
            });
        } finally {
            release();
        }
    }

    async showTime() {
        const release = await this.mutex.acquire();
        try {
            console.log(`現在の時刻: ${this.clock}`);
        } finally {
            release();
        }
    }
}

const main = async () => {
    const processNumber = parseInt(process.argv[2], 10);
    if (!PROCESS_PORT[processNumber]) {
        console.error('プロセスナンバーが割り当てられていません。');
        process.exit(1);
    }

    const lamportClock = new LamportClock();
    lamportClock.receiveThread(processNumber);

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    const prompt = () => {
        rl.question(`\n=== 時刻${lamportClock.clock} ===\n発生させるイベントを入力 (c: 計算イベント, s: メッセージ送信, t: 現在の時刻を表示): `, async (cmd) => {
            if (cmd === 'c') {
                await lamportClock.calcEvent();
            } else if (cmd === 's') {
                rl.question('プロセスナンバー: ', async (processNum) => {
                    await lamportClock.sendMessage(parseInt(processNum, 10));
                });
            } else if (cmd === 't') {
                await lamportClock.showTime();
            } else {
                console.log('再入力');
            }
            prompt();
        });
    };

    prompt();
};

main();
